strawberry/promise/async_.py

```diff
@@ -0,0 +1,159 @@
+# Copy of the Promise library (https://github.com/syrusakbary/promise) with some
+# modifications.
+#
+# Promise is licensed under the terms of the MIT license, reproduced below.
+#
+# = = = = =
+#
+# The MIT License (MIT)
+#
+# Copyright (c) 2016 Syrus Akbary
+#
+# Permission is hereby granted, free of charge, to any person obtaining a copy
+# of this software and associated documentation files (the "Software"), to deal
+# in the Software without restriction, including without limitation the rights
+# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+# copies of the Software, and to permit persons to whom the Software is
+# furnished to do so, subject to the following conditions:
+#
+# The above copyright notice and this permission notice shall be included in all
+# copies or substantial portions of the Software.
+#
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+# SOFTWARE.
+
+# Based on https://github.com/petkaantonov/bluebird/blob/master/src/promise.js
+from collections import deque
+from threading import local
+from typing import Any, Callable, Optional
+
+
+if False:
```

Comment:
What's the purpose of this?

Reply:
This comes from the upstream source of this file. I think it's supposed to be like `if TYPE_CHECKING` but I don't know why they didn't do that instead.

Reply:
maybe because they needed support for python 2 (and < 3.5) without having to install any other library :)